plugins/uefi-capsule/fu-uefi-bootmgr.c

```diff
@@ -249,6 +249,147 @@ fu_uefi_setup_bootnext_with_loadopt(FuEfiLoadOption *loadopt, GError **error)
 	return TRUE;
 }
 
+static GHashTable *
+fu_uefi_bootmgr_parse_csv(const gchar *data, size_t size, GError **error)
+{
+	g_auto(GStrv) lines = NULL;
+	g_autoptr(GHashTable) hash = NULL;
+	g_autofree gchar *nt_data = NULL;
+
+	hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
+
+	/* null terminate the file */
+	nt_data = g_strndup(data, size);
+
+	lines = g_strsplit(nt_data, "\n", -1);
+
+	for (guint i = 0; lines[i] != NULL; i++) {
+		gchar *endptr = NULL;
+		g_auto(GStrv) columns = NULL;
+		gint64 version;
+		g_autofree gint64 *version_bucket = NULL;
+
+		if ((lines[i][0] == '\0') && (lines[i + 1] == NULL)) {
+			/* last line may be empty */
+			break;
+		}
+
+		columns = g_strsplit(lines[i], ",", 3);
+		if (g_strv_length(columns) < 2) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid CSV data: not enough columns");
+			return NULL;
+		}
+
+		version = g_ascii_strtoll(columns[1], &endptr, 10);
+		if ((endptr == columns[1]) || (*endptr != '\0')) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid version in sbat level for entry %s",
+				    columns[0]);
+		}
+		version_bucket = g_malloc(sizeof(gint64));
+		g_assert(version_bucket != NULL);
+		*version_bucket = version;
+		g_hash_table_insert(hash, g_strdup(columns[0]), g_steal_pointer(&version_bucket));
+	}
+
+	return g_steal_pointer(&hash);
+}
+
+static gboolean
+fu_uefi_bootmgr_shim_is_safe(const gchar *source_shim, GError **error)
+{
+	g_autofree gchar *sbatlevel_path = NULL;
+	g_autofree gchar *sbatlevel_content = NULL;
+	gsize sbatlevel_size = 0;
+	g_autofree gchar *current_sbatlevel = NULL;
+	gsize current_sbatlevel_size = 0;
+	g_autoptr(GHashTable) shim_hash = NULL;
+	g_autoptr(GHashTable) current_hash = NULL;
+	GHashTableIter iter;
+	gpointer key, value;
+
+	sbatlevel_path = g_strdup_printf("%s.sbatlevel", source_shim);
+	if (!g_file_test(sbatlevel_path, G_FILE_TEST_EXISTS)) {
```

Comment:
Once you start natively parsing the shim I think this block can go; the checks will happen even in traditional distros.